webp: Allow unknown chunks at the end of the file

hanmertens opened the pull request:
Issue #1873 shows a number of real-world WebP images fail during decoding. For almost all of these, this is because they contain ["unknown chunks"](https://developers.google.com/speed/webp/docs/riff_container#unknown_chunks) at the end of the file. Those unknown chunks were reported as errors, while the specification says they should be ignored at the end of the file. This PR changes the decoder's behavior accordingly.

One of the images provided in #1873 still fails to decode after this PR. This is an animation file, so it might be related to that the standard also specifies that unknown chunks should be ignored at the end of animation "ANMF" chunks, which I don't think is currently implemented. However, running `webpmux -info` on this file resulted in `Failed to create mux object from file chunk_header_invalid/Ataturk-Bayrak.webp.`, so it may also just be a malformed file.

I license past and future contributions under the dual MIT/Apache-2.0 license,
allowing licensees to choose either at their option.

fintelia commented on src/codecs/webp/extended.rs:
```diff
@@ -536,6 +536,23 @@ struct AnimatedFrame {
+        match read_chunk(r) {
+            Ok(None) => return Ok(None),
+            Ok(chunk) => return unknown_chunk.and(Ok(chunk)),
+            Err(err @ ImageError::Decoding(_)) => unknown_chunk = unknown_chunk.and(Err(err)),
```
Isn't this going to catch all chunks that failed to decode, regardless of whether they're of known or unknown chunk types?

hanmertens replied:
Having an unknown chunk type is, at least currently, the only way a chunk can fail to decode with `ImageError::Decoding`, since the chunk's contents aren't decoded at this point.

hanmertens replied:
Probably best if it's split in a separate function (like `read_chunk_inner`) to prevent changing `read_chunk`'s signature that's also used elsewhere, but it should be doable. I'll have a look later.

hanmertens commented:
I've split off a function to read the chunk header to make sure the decoding error corresponds to an unknown chunk type.

I also found a fix for #1874: the code for removing the padding byte of a chunk didn't actually remove that byte. Since it's a relatively minor change, I added it to this PR instead of making a new one.

fintelia commented:
Thanks!
